## Algorithm

Differences in correlations between traits within an inbred population are 
determined at each genetic marker. Phenotypes are assigned to genotype groups 
and a single phenotype is used to scan all other phenotypes for a loss or gain 
of correlation. The likelihood profiles (~ QTL profiles) of this 'loss of 
correlation' measurement shows a very high degree of overlap with classical 
QTL profiles. However additional information is available from the phenotype x 
phenotype interactions. With the right dataset (ideally a combination of: 
classical phenotypes, protein abundance and gene expression) CTL shows the 
genetic wiring of the classical phenotypes and identify key players in the 
genetic / protein network underlying classical phenotypes using QTL and CTL 
information.

For more information on how-to calculate CTLs on N-genotypes, please read 
Chapter 4 of [my Thesis](http://hdl.handle.net/11370/29cf3cc5-6596-4d5a-a87c-490b31676a96)