%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Danny Arends
% The qcl mapping routine
%\VignetteIndexEntry{Article}
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[12pt]{article}

\usepackage{times}
\usepackage{amsmath}
\usepackage{url}
\usepackage{color}

% revise margins
\setlength{\headheight}{0.0in}
\setlength{\topmargin}{-0.25in}
\setlength{\headsep}{0.0in}
\setlength{\textheight}{9.00in}
\setlength{\footskip}{0.5in}
\setlength{\oddsidemargin}{0in}
\setlength{\evensidemargin}{0in}
\setlength{\textwidth}{6.5in}

\setlength{\parskip}{6pt}
\setlength{\parindent}{0pt}

\newcommand{\code}{\texttt}
\newcommand{\lod}{\text{LOD}}

% make "Figure" within figure captions a small font
\renewcommand{\figurename}{\small Figure}

\begin{document}

\SweaveOpts{prefix.string=fig}
\setkeys{Gin}{width=\textwidth} %% <- change width of figures
% Try to get the R code from running into the margin
<<echo=FALSE>>=
options(width=87, digits=3, scipen=4)
@ 

<<echo=FALSE,results=hide>>=
  require(qcl)
  set.seed(61777369)
  data(ath.metabolites) # Arabidopsis Thaliana dataset
  qcl_result <- QCLscan(ath.metab$genotypes, ath.metab$phenotypes)
@

% Change S input/output font size
\DefineVerbatimEnvironment{Sinput}{Verbatim}{fontsize=\footnotesize, baselinestretch=0.75, formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{fontsize=\footnotesize, baselinestretch=0.75, formatcom = {\color[rgb]{0.56, 0, 0}}}

\begin{center}
\vspace*{1in}
\textbf{\Large QCL: Added value of differential correlation }
 
\bigskip \bigskip 
\bigskip \bigskip 
 
{\large Danny Arends$^{1,2}$ and Ritsert C. Jansen$^{1}$} 

\bigskip \bigskip

{\color{blue} \tt \bfseries $^{1}$ Groningen Bioinformatics Centre - University of Groningen\\
$^{2}$ Genomics Coordination Centre - University Medical Centre Groningen\\
version: \# 0.0.1\\[24pt]
received on XX XXX 201X\\
accepted on XX XXX 201X\\
}

\end{center}

\bigskip \bigskip

\textbf{\sffamily Abstract}: In this paper we present Quanitative Correlation Locus 
(QCL) mapping, a novel approach to detect genetic regulation of phenotypes in 
recombinant inbred line populations (RIL) or other experimental crosses. QCL mapping 
allows association of phenotype correlation differences to genetic markers, in almost 
the same way as QTL mapping associates genetic markers with differences in expression.
We use QCL mapping to reconstruct underlying genetic architecture, and provide insight 
into regulation of (gene) expression. QCL information in combination with classical QTL 
information leads to a more detailed insight into the genetic architecture underlying 
complex traits. 
In two example we reconstruct part of the genetic architecture in S. Cerevisiae, another 
example shows the reconstruction of the chain elongation pathway in A. Thaliana. In both 
examples we are able to map the observe differences in correlation to genetic markers, 
and show the biological relevance of QCL peeks and hotspots in regulatory network 
reconstruction.

\vfill

\noindent Corresponding Author: Danny Arends \verb|Danny.Arends@Gmail.com|
\noindent Associate Editor: XXXX XXXX
\newpage

\textbf{\sffamily Introduction}

Genetical genomics experiments have shown that gene expression variation can be mapped to 
genotypic variation (Jansen and Nap, 2001). Quantitative Trait Locus (QTL) mapping of a 
gene identifies regions in the genome for which different genotypes lead to differential 
expression level of the gene. This QTL mapping can be done on all molecular levels including, 
gene expression (eQTL), protein abundance (pQTL) and metabolites (mQTL).

Recently different methods have been proposed that allow studying, in addition to the genes 
differentially expressed between conditions, the genes that are differentially coexpressed 
between conditions (Tesson et al. 2010, Kompass et al. 2011). Such differential co-expression 
is taken as another level of information from genomic data. This level of information can for 
example be used to expose transcriptional network rewiring (Langfelder et al. 2008). Differential 
co-expression has been observed in studies of different diseases, including between healthy and 
cancer tissue studies. In parallel, correlations between phenotypes are studied for various 
reasons e.g. discovering and detecting batch effects, marker assisted breeding and causal 
inference. The importance of identifying differences in correlation has been increasingly 
recognized and this led to the development of this algorithm to perform QCL analysis in 
experimental crosses.

We proposed to map not differential co-expression but differential correlation to genetic 
variation, i.e. to identify regions in the genome for which one genotype value leads to 
correlated expression between a start gene and all other genes, exploiting the fact that any 
cis-QTL will induce variation which will in turn lead to correlation differences among 
individuals carrying the other genotype. We merge information comming from QCL and QTL to reveal 
underlying genetic architecture, which would remain hidden when using only classical QTL information.

We urge all QTL researchers to combine QCL analysis with existing QTL analysis to expose 
transcriptional networks involved in the regulation of complex phenotypes. For this we 
provide an R package with a reference implementation  of the QCL algorithm

\bigskip 

\textbf{\sffamily Methodology}
\nopagebreak

We show two examples in which we reconstruct the genetic architecture using QCL mapping. 
Both datasets are available inside the R package, as well as the code to analyse these 
two datasets.

\emph{\sffamily Dataset Example 1. A. Thaliana}

Metabolite data from an A. Thaliana Landsberg erecta (Ler) and Cape Verde Islands (Cvi) 
dataset from R/qtl, 24 characterized metabolites which are the products of chain elongation. 
There are 162	recombinant inbred lines genotyped at 117 markers on 5 chromosomes. Additional 
details about the experimental settings are available in the original paper (Keurentjes et al. 2006,
Alonso-Blanco et al. 1998)

\emph{\sffamily Dataset Example 2. S. Cerevisiae}

S. Cerevisiae dataset from the gene expression omnibus (GEO) (GDS1115 / GDS1116). This 
dataset contains expression profiles of parental strains and haploid progenies from a 
cross of strain BY4716 and the wild wine strain RM11-1a. Data was collected for 4482 
probes. Genotype data for the 109 progenies was available at 282 genetic markers 
obtained by contacting the authors of the original paper. Expression of probes was 
measured under growth conditions containing 1\% ethanol. Additional details about the 
experimental settings are available in the original paper (Brem et al. 2005).

\emph{\sffamily Preprocessing}

Probes showing little to no variance (< 0.01) were removed from the analysis. Batch 
effects due to different hybridization dates were estimated and removed from the data 
by using a mean centering approach (Group means were estimated and corrected).

\textbf{\sffamily Algorithm: QCL mapping}

Using recombinant inbred lines we introduce natural variation into a population 
detectable as cis-QTL in a genome wide QTL scan. The effect of any such genetic 
variation is transfered downstream to interaction partners. When this leads to 
changes in mean expression of one of the partners we can also detect this using 
classical QTL methods (trans/distant QTL). 

However QTL mapping suffers from two major drawbacks (p1) the interaction partners 
of genes are obscured by many phenotypes co-localizing at a single genetic marker, and 
(p2) due to buffering the effect of any variation will 'fade' quickly and not be 
detectable as the difference in mean phenotype in the population does not change 
significantly.

QCL mapping however uses this buffering to detect interaction partners. The hypothesis 
is that when we split the population based on genotype at the cis-marker, related phenotypes 
will loosing correlation with the gene under investigation. This allows us to not only 
only identify its interaction partners. QCL mapping is thus not suffering from (p1) many 
phenotypes co-localizing). Additionally it is more sensitive to detect variation caused by 
regulators genes because variation of these genes causes many 'downstream' phenotypes 
to show differential correlation with the inducing phenotype, because QCL relies on 
correlation changes as opposed to differences in phenotype mean QCL is unaffected by (p2).

\emph{\sffamily Choosing a start/seed probe (Optional)}

Choose a phenotype to subject to QCL analysis, this can be your favourite gene, 
protein, metabolite or classical phenotype.

\emph{\sffamily Single phenotype QCL}

Using genotype data we can split individuals at each marker in two groups (A or B).
At each marker calculate the correlation between our phenotype and all other 
phenotypes in groups A and B. Collect all the result vectors into two matrices A 
and B, and calculate the QCL matrix by taking the squared difference between matrix A 
and B.

\emph{\sffamily Significance testing}

The threshold for a significant difference in QCL is assessed by using permutations, in 
each round the link between genotype and phenotype is broken, by redistributing at random 
genotypes amongst individuals not allowing for duplicates. The QCL on these new 
individuals allows for estimation of the underlying null-distribution of QCL scores. 
See figures XX, and XX showing the score distribution in permuted data.

\emph{\sffamily Creating single trait QCL profiles}

After permutation single trait QCL profiles can be obtained by setting a qcl significance 
threshold. This threshold defines when a change in correlation is significant, and at 
each marker we can plot the count of number the other phenotypes that show a significant 
difference in correlation. Overplotting chromosome, QTL information, and phenotype location 
(if applicable) produces the single QCL plots seen in figures XX, and XX.

\begin{figure}
\centering
<<singletrait, fig=TRUE, echo=FALSE, height=3.5>>=
plot(qcl_result, 2, do.legend=FALSE)
@ 
\caption{Single trait QCL profile (A. Thaliana).\label{fig:singletrait}}
\end{figure}

\emph{\sffamily Step 3 Network scan for QCL}

If we find any significant QCL information we can try to reconstruct the interaction 
network of the phenotype. We first do a single phenotype QCL scan (step 2), then 
using a permutation determin a qcl threshold, and select phenotypes which show a 
significant change in QCL.

Two things can happen: 
1) No genes show a significant change in QCL, which means we 
cannot find any interaction partners for this phenotype, or 
2) This genes has an QCL with other genes, then the algorithm adds 
these genes onto a list to scan and repeats the single phenotype 
QCL scan (step 2) for these phenotypes (not allowing for duplicate 
scans). This is repeated until no more significant QCL are found or 
a user defined search depth is reached.

\emph{\sffamily Visualizing the Phenotypes x Marker relations}

When plotting multiple QCL profiles as a heatmap, multiple QCL are seen mapping to the 
same genetic location. At these locations, we expect to find one (or more) of the genes 
to show a cis-QTL, this is because a QTL can induce a QCL, because variation of expression 
in the phenotype is causing other phenotypes to become uncorrelated, when assigned to 
groups determined by the QTL marker. This phenotype (or phenotypes) introduced genetic 
variation, which is seen also (and more pronounced) in a QTL scan. 

\emph{\sffamily Visualizing the Phenotypes x Phenotypes relations}

The additional information obtained by QCL mapping is summarized in the 
Phenotypes x Phenotypes matrix. This shows the summed evidence of interaction 
between the phenotypes showing QCL transmitted through all the markers and the phenotype scanned.

\emph{\sffamily Network reconstruction}

So how do we build up gene interaction networks, This looks complicated but the information 
is implicitely stored in the Phenotype x Marker and Phenotype x Phentype relations matrices. 
By using network scanning we only identify part of the network in which our phenotype is 
active and closely related interaction partners. To identify a pathway or main regulatory phenotypes 
controlling our phenotype we need to combine this information.

\bigskip
\textbf{\sffamily Results}
\nopagebreak

Global overview of QCL scan after using a full scan for QCL
<<echo=FALSE>>=
  qcl_result
@ 
Single results from a QCL mapping, shows the pattern at the highest threshold.
<<echo=FALSE>>=
  qcl_result[[2]]
@ 

\emph{\sffamily The Phenotypes x Marker relations}

Phenotype x Marker image plot, this is 'comparable' to a QTL scan on mutliple traits.

\begin{figure}[ht]
\centering
<<p2mmatrix, fig=TRUE, echo=FALSE, height=3.5>>=
par(mar=c(3,8,1.1,1.1))
image(qcl_result, 0.5, against="markers",do.grid=FALSE)
@ 
\caption{Phenotypes x Markers matrix (A. Thaliana).\label{fig:p2mmatrix}}
\end{figure}

\emph{\sffamily The Phenotypes x Phenotypes relations}

Phenotype x Phenotype image plot, additional information obtained by QCL mapping, 
which can be used to reconstruct the underlying genetic architecture.
\begin{figure}[ht]
\centering
<<p2pmatrix, fig=TRUE, echo=FALSE, height=3.5>>=
par(mar=c(3,8,1.1,1.1))
image(qcl_result, 0.5, against="phenotypes")
@ 
\caption{Phenotypes x Phenotypes matrix (A. Thaliana).\label{fig:p2pmatrix}}
\end{figure}

\bigskip
\textbf{\sffamily Discussion}
\nopagebreak

Microarray experiments have allowed the identification of genes differentially 
expressed between various conditions, or groups of genes coexpressed across 
samples. 

Many genetical genomics screens have been carried out to map differential 
expression to genetic variation. Screening the genome for genotype combinations 
for which there is differential expression has been a successful strategy when 
mapping Mendelian traits. However results on complex traits have been less 
then satisfactory with only small amounts of heritability explained by 
individuals loci. 
QCL mapping however 'borrows' information from each phenotype in the analysis set, 
and enables extraction of phenotype to marker to phenotype networks. 
Furthermore QCLs seem to 'transfer' from phenotype to phenotypes in a stabile 
fashion, which gives greater detail and confidence when reconstructing genetic 
architecture. 
It can provides better insight in the missing of heritability of variation, we 
hypothesize that changes in variation are affecting (by loss of) correlation in 
the normally correlated phenotypes leading to an amplification of this variation, 
by distrupting the network environment of which endo-phenotypes are part of.

\newpage
\textbf{\sffamily Bibliography}

Langfelder P, Horvath S (2008) WGCNA: an R package for weighted correlation network 
analysis. \emph{BMC Bioinformatics}, 9:559

Jansen R. C, and Nap J. P. (2001) Genetical Genomics: the added value from seggregation. 
\emph{Trends Genet}, 17(7), 388-391.

Tesson, B. M., Breitling R. and Jansen, R. C. (2010). DiffCoEx: a simple and sensitive method 
to find differentially coexpressed gene modules. \emph{BMC Bioinformatics}, 11,497-506.

Kompass, K. and Witte, J. (2011) - Co-regulatory expression quantitative trait loci mapping: 
method and application to endometrial cancer. \emph{BMC Medical Genomics}, 4, 6.

Broman, K.W., Wu, H., Sen, S. and Churchill, G.A. (2003) - R/qtl: QTL mapping in experimental 
crosses. \emph{Bioinformatics}, 19, 889-890.

Arends, D., Prins, P., Jansen, R.C. and Broman, K.W. (2010) - R/qtl: high-throughput multiple 
QTL mapping. \emph{Bioinformatics}, 26, 2990-2992.

R Development Core Team (2010) - R: A Language and Environment for Statistical Computing. 
http://www.R-project.org, ISBN: 3-900051-07-0

Edgar R., Domrachev M., Lash A.E. (2002) - Gene Expression Omnibus: NCBI gene expression and 
hybridization array data repository. \emph{Nucleic Acids Res.}, 30,207-210.

Keurentjes, J. J., Fu, J., de Vos, C. H., Lommen, A., Hall, R. D., Bino, R. J., van der Plas, 
L. H., Jansen, R. C., Vreugdenhil, D., Koornneef, M. (2006) - The genetics of plant metabolism. 
\emph{Nature Genetics.}, 38(7), 842-849.

Alonso-Blanco, C., Peeters, A. J., Koornneef, M., Lister, C., Dean, C., van den Bosch, N., Pot, 
J., Kuiper, M. T. (1998) - Development of an AFLP based linkage map of Ler, Col and Cvi 
Arabidopsis thaliana ecotypes and construction of a Ler/Cvi recombinant inbred line population. 
\emph{Plant J.}, 14, 259-271.

Brem RB, Storey JD, Whittle J, Kruglyak L. (2005) - Genetic interactions between polymorphisms 
that affect gene expression in yeast. \emph{Nature}, 436(7051),701-703.

\newpage
{\small
\noindent \textbf{\sffamily Package qcl function overview} \\
\noindent \begin{tabular}{ll} 
\hspace*{25mm} & \hspace*{103mm} \\ \hline
QCLscan & Genome wide scan for QCL \\
QCLscan.network & Scan for QCL using start gene(s) \\
QCLpermute & Significance thresholds for QCL by permutation \\
plot.QCL & Plot QCL profiles at increasing cut-offs \\
image.QCL & Image a multiple phenotype scan \\
\hline
\end{tabular}
}

\end{document}
